src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -14,211 +24,3020 @@ namespace Microsoft.CodeAnalysis.CSharp
             {
-                Optional<TLocalState> oldTryState = _tryState;
-                _tryState = AllBitsSet();
-                VisitTryBlock(tryBlock, node, ref tryState);
-                var tempTryStateValue = _tryState.Value;
-                IntersectWith(ref tryState, ref tempTryStateValue);
-                if (oldTryState.HasValue)
+                get
                 {
-                    var oldTryStateValue = oldTryState.Value;
-                    IntersectWith(ref oldTryStateValue, ref tempTryStateValue);
-                    oldTryState = oldTryStateValue;
+                    if (Branch == null) return null;
+                    switch (Branch.Kind)
+                    {
+                        case BoundKind.GotoStatement: return ((BoundGotoStatement)Branch).Label;
+                        case BoundKind.ConditionalGoto: return ((BoundConditionalGoto)Branch).Label;
+                        case BoundKind.BreakStatement: return ((BoundBreakStatement)Branch).Label;
+                        case BoundKind.ContinueStatement: return ((BoundContinueStatement)Branch).Label;
+                        case BoundKind.PatternSwitchLabel: return ((BoundPatternSwitchLabel)Branch).Label;
+                        default: return null;
+                    }
                 }
+            }
 
-                _tryState = oldTryState;
+            public PendingBranch(BoundNode branch, TLocalState state)
+            {
+                this.Branch = branch;
+                this.State = state.Clone();
             }
-            else
+        }
+
+        /// <summary>
+        /// Perform a single pass of flow analysis.  Note that after this pass,
+        /// this.backwardBranchChanged indicates if a further pass is required.
+        /// </summary>
+        protected virtual ImmutableArray<PendingBranch> Scan(ref bool badRegion)
+        {
+            var oldPending = SavePending();
+            Visit(methodMainNode);
+            this.Unsplit();
+            RestorePending(oldPending);
+            if (_trackRegions && regionPlace != RegionPlace.After) badRegion = true;
+            ImmutableArray<PendingBranch> result = RemoveReturns();
+            return result;
+        }
+
+        protected ImmutableArray<PendingBranch> Analyze(ref bool badRegion)
+        {
+            ImmutableArray<PendingBranch> returns;
+            do
             {
-                VisitTryBlock(tryBlock, node, ref tryState);
+                // the entry point of a method is assumed reachable
+                regionPlace = RegionPlace.Before;
+                this.State = TopState();
+                PendingBranches.Clear();
+                if (_trackExceptions) PendingBranches.Add(new PendingBranch(null, TopState()));
+                this.stateChangedAfterUse = false;
+                this.Diagnostics.Clear();
+                returns = this.Scan(ref badRegion);
             }
+            while (this.stateChangedAfterUse);
```

Comment:
is it necessary for stateChangedAfterUse to have to be instance state that is mutated?  would it be possible for it to be more of an out-var of Scan, along with bad-region/returns?